specs/data_structures.md

```diff
@@ -161,23 +174,88 @@ https://developers.google.com/protocol-buffers/docs/proto3
 
 # Hashing
 
-https://en.wikipedia.org/wiki/SHA-3
-
-https://godoc.org/golang.org/x/crypto/sha3
+All protocol-level hashing is done using [Keccak-256](https://keccak.team/keccak.html), and not SHA3-256 ([FIPS 202](https://keccak.team/specifications.html#FIPS_202)). This is to enable compatibility with [Ethereum](https://ethereum.org)'s EVM. Keccak-256 outputs a digest that is 256 bits (i.e. 32 bytes) long.
 
-https://docs.rs/sha3
+Libraries implementing Keccak-256 are available in Go (https://godoc.org/golang.org/x/crypto/sha3) and Rust (https://docs.rs/sha3).
 
 # Signing
 
 
 
-# Merkle Tree
+# Merkle Trees
+
+Merkle trees are used to authenticate various pieces of data across the LazyLedger stack, including transactions, messages, the validator set, etc. This section provides an overview of the different tree types used, and specifies how to construct them.
+
+## Binary Merkle Tree
+
+Binary Merkle trees are constructed in the usual fashion, namely:
+
+For leaf node of message `m`:
+```C++
+v = h(serialize(m))
+```
+
+For internal node with children `l` and `r`:
+```C++
+v = h(l, r) = h(l.v, r.v)
+```
```

Comment:
Isn't it essential that leaf nodes vs internal nodes get hashed differently? Usually that happens with a prefix. e.g. https://github.com/tendermint/tendermint/blob/48f073d796ca4d1063b5f02c5d1b35c2c7e77afc/crypto/merkle/hash.go#L8-L21
Some discussions about this here: https://github.com/tendermint/tendermint/issues/1892

Reply:
Note that the flaghasher in @musalbas prototype also differentiates between leaf (and other) data: https://github.com/LazyLedger/lazyledger-prototype/blob/60faf6af245830ebe25a4963d587e96a467d2eb1/flaghasher.go#L71-L76 
(it additionally has a `codedFlag` if in codec mode). And the leaf hash is computed prefixed with `0x00` (using that flaghasher): https://github.com/LazyLedger/lazyledger-prototype/blob/60faf6af245830ebe25a4963d587e96a467d2eb1/merkletree_exports.go#L20-L22